framework/auth/views.py

```diff
@@ -202,92 +202,160 @@ def forgot_password_post(auth, **kwargs):
+def login_and_register_handler(auth, login=True, campaign=None, service_url=None, logout=None):
     """
-    This view serves as the entry point for OSF login and campaign login.
-    HTTP Method: GET
+    Non-view helper to handle `login` and `register` requests.
 
-        GET '/login/' without any query parameter:
-            redirect to CAS login page with dashboard as target service
+    :param auth: the auth context
+    :param login: `True` if `GET /login`, `False` if `GET /register`
+    :param campaign: a target campaign defined in `auth.campaigns`
+    :param service_url: the service url for CAS login or redirect url for OSF
+    :param logout: used only for `claim_user_registered`
+    :return: data object that contains actions for `auth_register` and `auth_login`
+    :raises: http.BAD_REQUEST
+    """
 
-        GET '/login/?logout=true
-            log user out and redirect to CAS login page with redirect_url or next_url as target service
+    # set target page to `dashboard` if no service url is specified
+    if not service_url:
+        service_url = web_url_for('dashboard', _absolute=True)
 
-        GET '/login/?campaign=instituion:
-            if user is logged in, redirect to 'dashboard'
-            show institution login
+    # Only allow redirects which are relative root or full domain. Disallows external redirects.
+    if not validate_service_url(service_url):
+        raise HTTPError(http.BAD_REQUEST)
 
-        GET '/login/?campaign=prereg:
-            if user is logged in, redirect to prereg home page
-            else show sign up page and notify user to sign in, set next to prereg home page
+    data = {
+        'status_code': http.FOUND if login else http.OK,
+        'redirect_url': service_url,
+        'campaign': None,
+        'must_login_warning': False,
+        'service_url': service_url,
+    }
 
-        GET '/login/?next=next_url:
-            if user is logged in, redirect to next_url
-            else redirect to CAS login page with next_url as target service
-    """
+    if campaign:
+        if validate_campaign(campaign):
+            # GET `/register` or '/login` with `campaign=institution`
+            # unlike other campaigns, institution login serves as an alternative for authentication
+            if campaign == 'institution':
+                data['status_code'] = http.FOUND
+                data['redirect_url'] = cas.get_login_url(service_url, campaign='institution')
+            # For all other non-institution campaigns
+            else:
+                # `GET /login?campaign=...`
+                if login:
+                    service_url = data['redirect_url'] = web_url_for('auth_register', campaign=campaign)
+                # `GET /register?campaign=...`
+                else:
+                    data['campaign'] = campaign
+                    data['must_login_warning'] = True
+                    service_url = data['service_url'] = campaigns.campaign_url_for(campaign)
```

Comment:
Is `service_url` or `data['service_url']` necessary? The one conditional branch in which it's accessed, it looks like `data['redirect_url']` isn't accessed.